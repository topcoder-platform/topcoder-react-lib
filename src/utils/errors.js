/**
 * @module errors
 * @desc Utility functions for the standard error handling.
 * @todo They will be moved to `topcoder-react-utils` soon.
 */

import { isomorphy } from 'topcoder-react-utils';
import actions from '../actions/errors';

export const ERROR_ICON_TYPES = {
  NETWORK: 'network',
  API: 'api',
};

let store; // Will be set only when rendering client-side

export function setErrorsStore(s) {
  store = s;
}

/**
 * The function behaves similarly to javascript alert()
 * it will show a modal error diaglog with styling until the user clicks OK.
 */
export function fireErrorMessage(title, details, support) {
  if (isomorphy.isClientSide() && store) {
    setImmediate(() => {
      store.dispatch(actions.errors.newError(title, details, support));
    });
  }
}

/**
 * clear all error icons
 */
export function clearAllErrorIcons() {
  if (isomorphy.isClientSide() && store) {
    setImmediate(() => {
      store.dispatch(actions.errors.clearAllErrorIcons());
    });
  }
}

/**
 * add error message to error icon in bottom left in the screen.
 * @param id  id of error icon type (ERROR_ICON_TYPES.NETWORK or ERROR_ICON_TYPES.API)
 * @param title icon hover title
 * @param message icon hover message
 */
export function setErrorIcon(id, title, message) {
  if (isomorphy.isClientSide() && store) {
    setImmediate(() => {
      store.dispatch(actions.errors.setErrorIcon(id, title, message));
    });
  }
}

/**
 * clear all error message for an error icon.
 * @param id  id of error icon type to clear
 */
export function clearErrorIcon(id) {
  if (isomorphy.isClientSide() && store) {
    setImmediate(() => {
      store.dispatch(actions.errors.clearErrorIcon(id));
    });
  }
}
