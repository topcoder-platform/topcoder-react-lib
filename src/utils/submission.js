/**
 * Various submissions functions.
 */

/**
 * Get provisional score of submission
 * @param submission
 */
export function getProvisionalScore(submission) {
  const { submissions: subs } = submission;
  if (!subs || subs.length === 0) {
    return 0;
  }
  const { provisionalScore } = subs[0];
  if (!provisionalScore || provisionalScore < 0) {
    return 0;
  }
  return provisionalScore;
}

/**
 * Get final score of submission
 * @param submission
 */
export function getFinalScore(submission) {
  const { submissions: subs } = submission;
  if (!subs || subs.length === 0) {
    return 0;
  }
  const { finalScore } = subs[0];
  if (!finalScore || finalScore < 0) {
    return 0;
  }
  return finalScore;
}
