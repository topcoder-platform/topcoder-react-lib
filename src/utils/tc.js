/**
 * @module tc
 * @desc Collection of small Topcoder-related functions.
 * @todo More TC-related utils should be moved here from Community-app.
 */

/**
 * Codes of the Topcoder communities.
 */
/* TODO: These are originally motivated by Topcoder API v2. Topcoder API v3
 * uses upper-case literals to encode the tracks. At some point, we should
 * update it in this code as well! */
export const COMPETITION_TRACKS = {
  DATA_SCIENCE: 'data_science',
  DESIGN: 'design',
  DEVELOP: 'develop',
};

/**
 * Review Opportunity types
 */
export const REVIEW_OPPORTUNITY_TYPES = {
  'Contest Review': 'Review',
  'Spec Review': 'Specification Review',
  'Iterative Review': 'Iterative Review',
};

/**
 * Gets payload from a standard success response from TC API; or throws
 * an error in case of a failure response.
 * @param {Object} res
 * @param {Boolean} shouldThrowError should throw error if request fail
 * @return {Promise} Resolves to the payload.
 */
export async function getApiResponsePayload(res, shouldThrowError = true) {
  if (!res.ok) {
    if (shouldThrowError) {
      throw new Error(res.statusText);
    } else {
      return null;
    }
  }
  const x = (await res.json()).result;
  if ((!x.success)) {
    if (shouldThrowError) {
      throw new Error(x.content);
    } else {
      return null;
    }
  }
  return x.content;
}

/**
 * Gets payload from a standard success response from TC LOOKER API; or throws
 * an error in case of a failure response.
 * @param {Object} res
 * @return {Promise} Resolves to the payload.
 */
export async function getLookerApiResponsePayload(res) {
  const resJson = await res.json();
  if (Array.isArray(resJson)) {
    return {
      res: resJson,
      error: false,
    };
  }

  const x = resJson.result;
  return {
    res: x.content,
    error: !x.success,
    status: x.status,
  };
}
