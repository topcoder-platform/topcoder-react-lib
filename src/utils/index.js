/**
 * Export utils.
 */
import logger from './logger';
import * as tc from './tc';
import * as time from './time';
import * as mock from './mock';
import * as errors from './errors';
import * as filter from './challenge/filter';
import * as buckets from './challenge/buckets';
import * as sort from './challenge/sort';
import * as url from './url';
import * as submission from './submission';

const challenge = {
  filter,
  buckets,
  sort,
};

export {
  challenge,
  logger,
  tc,
  time,
  mock,
  errors,
  url,
  submission,
};
