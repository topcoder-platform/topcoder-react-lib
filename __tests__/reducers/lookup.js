import { mockAction } from 'utils/mock';

const tag = {
  domain: 'SKILLS',
  id: 251,
  name: 'Jekyll',
  status: 'APPROVED',
};

const mockActions = {
  lookup: {
    getSkillTagsInit: mockAction('LOOKUP/GET_SKILL_TAGS_INIT'),
    getSkillTagsDone: mockAction('LOOKUP/GET_SKILL_TAGS_DONE', [tag]),
    getSkillTagsDoneError: mockAction('LOOKUP/GET_SKILL_TAGS_DONE', null, 'Unknown error'),
  },
};
jest.setMock(require.resolve('actions/lookup'), mockActions);

const reducers = require('reducers/lookup');

let reducer;

function testReducer() {
  let state;

  test('Initial state', () => {
    state = reducer(undefined, {});
    expect(state).toMatchSnapshot();
  });

  test('Get skill tags', () => {
    state = reducer(state, mockActions.lookup.getSkillTagsInit());
    state = reducer(state, mockActions.lookup.getSkillTagsDone());
    expect(state).toMatchSnapshot();
  });

  test('Get skill tags error', () => {
    state = reducer(state, mockActions.lookup.getSkillTagsDoneError());
    expect(state).toMatchSnapshot();
  });
}

describe('Default reducer', () => {
  reducer = reducers.default;
  testReducer();
});

describe('Factory without server side rendering', () => {
  beforeAll((done) => {
    reducers.factory().then((res) => {
      reducer = res;
      done();
    });
  });
  testReducer();
});
