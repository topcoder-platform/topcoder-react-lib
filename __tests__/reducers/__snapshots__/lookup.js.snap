// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default reducer Get skill tags 1`] = `
Object {
  "loadingSkillTagsError": false,
  "skillTags": Array [
    Object {
      "domain": "SKILLS",
      "id": 251,
      "name": "Jekyll",
      "status": "APPROVED",
    },
  ],
}
`;

exports[`Default reducer Get skill tags error 1`] = `
Object {
  "loadingSkillTagsError": true,
  "skillTags": Array [
    Object {
      "domain": "SKILLS",
      "id": 251,
      "name": "Jekyll",
      "status": "APPROVED",
    },
  ],
}
`;

exports[`Default reducer Initial state 1`] = `
Object {
  "skillTags": Array [],
}
`;

exports[`Factory without server side rendering Get skill tags 1`] = `
Object {
  "loadingSkillTagsError": false,
  "skillTags": Array [
    Object {
      "domain": "SKILLS",
      "id": 251,
      "name": "Jekyll",
      "status": "APPROVED",
    },
  ],
}
`;

exports[`Factory without server side rendering Get skill tags error 1`] = `
Object {
  "loadingSkillTagsError": true,
  "skillTags": Array [
    Object {
      "domain": "SKILLS",
      "id": 251,
      "name": "Jekyll",
      "status": "APPROVED",
    },
  ],
}
`;

exports[`Factory without server side rendering Initial state 1`] = `
Object {
  "skillTags": Array [],
}
`;
